storage/storage_dest.go

```diff
@@ -609,8 +641,28 @@ func (s *storageImageDestination) commitLayer(index int, info addedLayerInfo, si
 			return err
 		}
 
-		// FIXME: what to do with the uncompressed digest?
-		diffOutput.UncompressedDigest = info.digest
+		man, err := manifest.FromBlob(s.manifest, manifest.GuessMIMEType(s.manifest))
+		if err != nil {
+			return fmt.Errorf("parsing manifest: %w", err)
+		}
+
+		cb, err := s.getConfigBlob(man.ConfigInfo())
+		if err != nil {
+			return err
+		}
+
+		// retrieve the expected uncompressed digest from the config blob.
+		configOCI := &imgspecv1.Image{}
+		if err := json.Unmarshal(cb, configOCI); err != nil {
+			return err
+		}
+		if index >= len(configOCI.RootFS.DiffIDs) {
+			return fmt.Errorf("index %d out of range for configOCI.RootFS.DiffIDs", index)
+		}
+
+		// let the storage layer know what was the original uncompressed layer.
+		diffOutput.UncompressedDigest = configOCI.RootFS.DiffIDs[index]
```

Comment:
Isn’t this an untrusted unverified value? What does storing that to c/storage do to the security model?

Reply:
It is not done for security reasons, only to validate that we generate the expected diffID.  If you see no value in doing it, I can drop this part.

In this model even computing the digest for the whole .tar file wouldn't add anything to the security model, since we don't use it.  We only use the TOC digest that was validated.  So there is no risk that a malicious image could pretend to have a different TOC (and risk reusing it with another image) than the one we've validated

Reply:
This was fixed by storing the untrusted value in a layer flag instead.